compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeResolver.java

```diff
@@ -0,0 +1,2212 @@
+import org.wso2.ballerinalang.compiler.tree.types.BLangValueType;
+import org.wso2.ballerinalang.compiler.util.BArrayState;
+import org.wso2.ballerinalang.compiler.util.CompilerContext;
+import org.wso2.ballerinalang.compiler.util.ImmutableTypeCloner;
+import org.wso2.ballerinalang.compiler.util.Name;
+import org.wso2.ballerinalang.compiler.util.Names;
+import org.wso2.ballerinalang.compiler.util.TypeTags;
+import org.wso2.ballerinalang.util.Flags;
+
+import java.util.ArrayList;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Stack;
+import java.util.stream.Collectors;
+
+import static org.ballerinalang.model.symbols.SymbolOrigin.BUILTIN;
+import static org.ballerinalang.model.symbols.SymbolOrigin.SOURCE;
+import static org.ballerinalang.model.symbols.SymbolOrigin.VIRTUAL;
+import static org.wso2.ballerinalang.compiler.util.Constants.INFERRED_ARRAY_INDICATOR;
+import static org.wso2.ballerinalang.compiler.util.Constants.OPEN_ARRAY_INDICATOR;
+
+/**
+ * Resolve the types of module-level-constructs.
+ *
+ * @since 2201.7.0
+ */
+
+public class TypeResolver {
+
+    private static final CompilerContext.Key<TypeResolver> TYPE_RESOLVER_KEY = new CompilerContext.Key<>();
+
+    private final SymbolTable symTable;
+    private final Names names;
+    private final SymbolResolver symResolver;
+    private final SymbolEnter symEnter;
+    private final BLangDiagnosticLog dlog;
+    private final Types types;
+    private int typePrecedence;
+    private final TypeParamAnalyzer typeParamAnalyzer;
+    private final ConstantTypeChecker constantTypeChecker;
+    private final ConstantTypeChecker.ResolveConstantExpressionType resolveConstantExpressionType;
+    private final EffectiveTypePopulator effectiveTypePopulator;
+    private BLangAnonymousModelHelper anonymousModelHelper;
+    private BLangMissingNodesHelper missingNodesHelper;
+
+    private List<BLangTypeDefinition> resolvingTypeDefinitions = new ArrayList<>();
+    private HashMap<BIntersectionType, BLangIntersectionTypeNode> intersectionTypeList;
+    public HashSet<BLangConstant> resolvedConstants = new HashSet<>();
+    private ArrayList<BLangConstant> resolvingConstants = new ArrayList<>();
+    private Stack<String> resolvingModuleDefs;
+    private HashSet<BLangClassDefinition> resolvedClassDef = new HashSet<>();
+    private Map<String, BLangNode> modTable = new LinkedHashMap<>();
+    private Map<String, BLangConstantValue> constantMap = new HashMap<>();
+    private HashSet<LocationData> unknownTypeRefs;
+    private SymbolEnv pkgEnv;
+    private int currentDepth;
+    private Stack<BType> resolvingTypes;
+    public HashSet<BStructureType> resolvingStructureTypes = new HashSet<>();
```

Comment:
We need to find a way to make this concurrent safe per CompilationContext. Something like AnalyzerData in visitor should work. Create an issue to track this.